File: connection.go
package unrealircd

import (
	"crypto/tls"
	"encoding/base64"
	"encoding/json"
	"errors"
	"net/http"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// Querier interface for RPC calls
type Querier interface {
	Query(method string, params interface{}, noWait bool) (interface{}, error)
}

// Connection represents a connection to the UnrealIRCd RPC server
type Connection struct {
	conn     *websocket.Conn
	mu       sync.Mutex
	errno    int
	err      error
	nextID   int
	idMu     sync.Mutex
}

// Options for connecting to the RPC server
type Options struct {
	Context   *http.Client
	TLSVerify bool
	Issuer    string
}

// NewConnection creates a new connection to the UnrealIRCd RPC server
func NewConnection(uri, apiLogin string, options *Options) (*Connection, error) {
	if options == nil {
		options = &Options{}
	}

	dialer := websocket.Dialer{
		HandshakeTimeout: 10 * time.Second,
	}

	if options.Context != nil {
		dialer.NetDialContext = options.Context.Transport.(*http.Transport).DialContext
	}

	if !options.TLSVerify {
		dialer.TLSClientConfig = &tls.Config{
			InsecureSkipVerify: true,
		}
	}

	header := http.Header{}
	header.Set("Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(apiLogin)))

	conn, _, err := dialer.Dial(uri, header)
	if err != nil {
		return nil, err
	}

	c := &Connection{
		conn:   conn,
		nextID: 1,
	}

	if options.Issuer != "" {
		// Set issuer asynchronously
		go c.Query("rpc.set_issuer", map[string]interface{}{"name": options.Issuer}, true)
	} else {
		// Ping-pong
		conn.WriteMessage(websocket.PingMessage, []byte{})
	}

	return c, nil
}

// Query sends a JSON-RPC request and waits for the response
func (c *Connection) Query(method string, params interface{}, noWait bool) (interface{}, error) {
	c.idMu.Lock()
	id := c.nextID
	c.nextID++
	c.idMu.Unlock()

	req := map[string]interface{}{
		"jsonrpc": "2.0",
		"method":  method,
		"params":  params,
		"id":      id,
	}

	data, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}

	c.mu.Lock()
	err = c.conn.WriteMessage(websocket.TextMessage, data)
	c.mu.Unlock()
	if err != nil {
		return nil, err
	}

	if noWait {
		return true, nil
	}

	timeout := time.After(10 * time.Second)
	for {
		select {
		case <-timeout:
			return nil, errors.New("RPC request timed out")
		default:
			c.mu.Lock()
			c.conn.SetReadDeadline(time.Now().Add(10 * time.Second))
			_, message, err := c.conn.ReadMessage()
			c.mu.Unlock()
			if err != nil {
				return nil, err
			}

			var resp map[string]interface{}
			if err := json.Unmarshal(message, &resp); err != nil {
				return nil, err
			}

			if respID, ok := resp["id"].(float64); ok && int(respID) == id {
				if result, ok := resp["result"]; ok {
					c.errno = 0
					c.err = nil
					return result, nil
				}
				if errObj, ok := resp["error"].(map[string]interface{}); ok {
					if code, ok := errObj["code"].(float64); ok {
						c.errno = int(code)
					}
					if msg, ok := errObj["message"].(string); ok {
						c.err = errors.New(msg)
					}
					return nil, c.err
				}
			}
			// Not our response, continue (for streaming events)
		}
	}
}

// EventLoop waits for the next event (used for log streaming)
func (c *Connection) EventLoop() (interface{}, error) {
	c.mu.Lock()
	c.conn.SetReadDeadline(time.Now().Add(2 * time.Second))
	_, message, err := c.conn.ReadMessage()
	c.mu.Unlock()
	if err != nil {
		if netErr, ok := err.(interface{ Timeout() bool }); ok && netErr.Timeout() {
			return nil, nil // timeout, can retry
		}
		return nil, err
	}

	var resp map[string]interface{}
	if err := json.Unmarshal(message, &resp); err != nil {
		return nil, err
	}

	if result, ok := resp["result"]; ok {
		c.errno = 0
		c.err = nil
		return result, nil
	}

	if errObj, ok := resp["error"].(map[string]interface{}); ok {
		if code, ok := errObj["code"].(float64); ok {
			c.errno = int(code)
		}
		if msg, ok := errObj["message"].(string); ok {
			c.err = errors.New(msg)
		}
		return nil, c.err
	}

	return nil, errors.New("Invalid JSON-RPC data from UnrealIRCd: not an error and not a result")
}

// Errno returns the last error code
func (c *Connection) Errno() int {
	return c.errno
}

// Error returns the last error
func (c *Connection) Error() error {
	return c.err
}

// Rpc returns the RPC handler
func (c *Connection) Rpc() *Rpc {
	return &Rpc{querier: c}
}

// Stats returns the stats handler
func (c *Connection) Stats() *Stats {
	return &Stats{querier: c}
}

// User returns the user handler
func (c *Connection) User() *User {
	return &User{querier: c}
}

// Channel returns the channel handler
func (c *Connection) Channel() *Channel {
	return &Channel{querier: c}
}

// ServerBan returns the server ban handler
func (c *Connection) ServerBan() *ServerBan {
	return &ServerBan{querier: c}
}

// Spamfilter returns the spamfilter handler
func (c *Connection) Spamfilter() *Spamfilter {
	return &Spamfilter{querier: c}
}

// NameBan returns the name ban handler
func (c *Connection) NameBan() *NameBan {
	return &NameBan{querier: c}
}

// Server returns the server handler
func (c *Connection) Server() *Server {
	return &Server{querier: c}
}

// Log returns the log handler
func (c *Connection) Log() *Log {
	return &Log{querier: c}
}

// ServerBanException returns the server ban exception handler
func (c *Connection) ServerBanException() *ServerBanException {
	return &ServerBanException{querier: c}
}

File: unrealircd-rpc-cs/UnrealIRCdRPC/Connection.cs
using System;
using System.Collections.Generic;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace UnrealIRCdRPC
{
    // Querier interface for RPC calls
    public interface IQuerier
    {
        Task<JsonElement?> QueryAsync(string method, object? parameters, bool noWait);
    }

    // Connection represents a connection to the UnrealIRCd RPC server
    public class Connection : IQuerier, IDisposable
    {
        private readonly ClientWebSocket _webSocket;
        private readonly SemaphoreSlim _sendLock = new SemaphoreSlim(1, 1);
        private int _errno;
        private Exception? _error;
        private int _nextId;
        private readonly object _idLock = new object();

        // Options for connecting to the RPC server
        public class Options
        {
            public bool TlsVerify { get; set; } = true;
            public string? Issuer { get; set; }
        }

        // Constructor
        private Connection(ClientWebSocket webSocket)
        {
            _webSocket = webSocket;
            _nextId = 1;
        }

        // NewConnection creates a new connection to the UnrealIRCd RPC server
        public static async Task<Connection> NewConnectionAsync(string uri, string apiLogin, Options? options = null, CancellationToken cancellationToken = default)
        {
            options ??= new Options();

            var webSocket = new ClientWebSocket();

            // Set up headers for authentication
            webSocket.Options.SetRequestHeader("Authorization", "Basic " + Convert.ToBase64String(Encoding.UTF8.GetBytes(apiLogin)));

            if (!options.TlsVerify)
            {
                // Note: In .NET, skipping TLS verification requires a custom handler, but for simplicity, we'll assume it's handled elsewhere
                // For production, you might need to use HttpClient with custom validation
            }

            var uriObj = new Uri(uri);
            await webSocket.ConnectAsync(uriObj, cancellationToken);

            var connection = new Connection(webSocket);

            if (!string.IsNullOrEmpty(options.Issuer))
            {
                // Set issuer asynchronously
                _ = connection.QueryAsync("rpc.set_issuer", new { name = options.Issuer }, true);
            }
            // Ping-pong is handled automatically by ClientWebSocket

            return connection;
        }

        // Query sends a JSON-RPC request and waits for the response
        public async Task<JsonElement?> QueryAsync(string method, object? parameters, bool noWait)
        {
            int id;
            lock (_idLock)
            {
                id = _nextId++;
            }

            var request = new
            {
                jsonrpc = "2.0",
                method,
                @params = parameters,
                id
            };

            var json = JsonSerializer.Serialize(request);
            var data = Encoding.UTF8.GetBytes(json);

            await _sendLock.WaitAsync();
            try
            {
                await _webSocket.SendAsync(new ArraySegment<byte>(data), WebSocketMessageType.Text, true, CancellationToken.None);
            }
            finally
            {
                _sendLock.Release();
            }

            if (noWait)
            {
                return null;
            }

            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(10));
            var buffer = new byte[4096];
            var segment = new ArraySegment<byte>(buffer);

            while (!cts.Token.IsCancellationRequested)
            {
                WebSocketReceiveResult result;
                await _sendLock.WaitAsync();
                try
                {
                    result = await _webSocket.ReceiveAsync(segment, cts.Token);
                }
                finally
                {
                    _sendLock.Release();
                }

                if (result.MessageType == WebSocketMessageType.Close)
                {
                    throw new Exception("WebSocket closed");
                }

                var message = Encoding.UTF8.GetString(buffer, 0, result.Count);
                var response = JsonSerializer.Deserialize<Dictionary<string, object>>(message);

                if (response != null && response.TryGetValue("id", out var respIdObj) && respIdObj is JsonElement respIdElement && respIdElement.TryGetInt32(out var respId) && respId == id)
                {
                    if (response.TryGetValue("result", out var resultObj))
                    {
                        _errno = 0;
                        _error = null;
                        return (JsonElement)resultObj;
                    }
                    if (response.TryGetValue("error", out var errorObj) && errorObj is JsonElement errorElement)
                    {
                        var errorDict = JsonSerializer.Deserialize<Dictionary<string, object>>(errorElement.GetRawText());
                        if (errorDict != null)
                        {
                            if (errorDict.TryGetValue("code", out var codeObj) && codeObj is JsonElement codeElement && codeElement.TryGetInt32(out var code))
                            {
                                _errno = code;
                            }
                            if (errorDict.TryGetValue("message", out var msgObj) && msgObj is JsonElement msgElement)
                            {
                                _error = new Exception(msgElement.GetString());
                            }
                        }
                        throw _error ?? new Exception("Unknown error");
                    }
                }
                // Not our response, continue (for streaming events)
            }

            throw new TimeoutException("RPC request timed out");
        }

        // EventLoop waits for the next event (used for log streaming)
        public async Task<object?> EventLoopAsync()
        {
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(2));
            var buffer = new byte[4096];
            var segment = new ArraySegment<byte>(buffer);

            WebSocketReceiveResult result;
            await _sendLock.WaitAsync();
            try
            {
                result = await _webSocket.ReceiveAsync(segment, cts.Token);
            }
            catch (OperationCanceledException)
            {
                return null; // timeout
            }
            finally
            {
                _sendLock.Release();
            }

            if (result.MessageType == WebSocketMessageType.Close)
            {
                throw new Exception("WebSocket closed");
            }

            var message = Encoding.UTF8.GetString(buffer, 0, result.Count);
            var response = JsonSerializer.Deserialize<Dictionary<string, object>>(message);

            if (response != null)
            {
                if (response.TryGetValue("result", out var resultObj))
                {
                    _errno = 0;
                    _error = null;
                    return resultObj;
                }
                if (response.TryGetValue("error", out var errorObj) && errorObj is JsonElement errorElement)
                {
                    var errorDict = JsonSerializer.Deserialize<Dictionary<string, object>>(errorElement.GetRawText());
                    if (errorDict != null)
                    {
                        if (errorDict.TryGetValue("code", out var codeObj) && codeObj is JsonElement codeElement && codeElement.TryGetInt32(out var code))
                        {
                            _errno = code;
                        }
                        if (errorDict.TryGetValue("message", out var msgObj) && msgObj is JsonElement msgElement)
                        {
                            _error = new Exception(msgElement.GetString());
                        }
                    }
                    throw _error ?? new Exception("Unknown error");
                }
            }

            throw new Exception("Invalid JSON-RPC data from UnrealIRCd: not an error and not a result");
        }

        // Errno returns the last error code
        public int Errno => _errno;

        // Error returns the last error
        public Exception? Error => _error;

        // Dispose
        public void Dispose()
        {
            _webSocket?.Dispose();
            _sendLock?.Dispose();
        }

        // Module handlers
        public Rpc Rpc() => new Rpc(this);
        public Stats Stats() => new Stats(this);
        public User User() => new User(this);
        public Channel Channel() => new Channel(this);
        public ServerBan ServerBan() => new ServerBan(this);
        public Spamfilter Spamfilter() => new Spamfilter(this);
        public NameBan NameBan() => new NameBan(this);
        public Server Server() => new Server(this);
        public Log Log() => new Log(this);
        public ServerBanException ServerBanException() => new ServerBanException(this);
    }
}